### PR TITLE
Moved the ownership of IterativeVerifier from sharding/copydb to Ferry

### PR DESCRIPTION
**Pull request by shuhaowu**
What is done 
----------

This PR moves the initialization and management of IterativeVerifier from the application level to the core library. This is required to interrupt and resume the IterativeVerifier as otherwise the application
layer will have to assume the responsibility of resuming the IterativeVerifier correctly, which is a non-trivial task.

This also removes the `.Tables` member on the DataIterator and made it an argument of `.Run`.

Also added tests in the ruby integration tests for the IterativeVerifier.

Implication for configurations
------------------------------

Since the verifiers will be initialized by the Ferry using configurations specified in Config. The Config interface is almost the same as the previous config, so a small change for sharding is needed:

Before for ghostferry-sharding:

```
{
 "IgnoredVerificationTables": [],
 "IgnoredVerificationColumns": [],
 "MaxExpectedVerifierDowntime": "1m",
 "VerifierIterationConcurrency": 4
}
```

After for ghostferry-sharding:

```
{
  "IterativeVerifierConfig": {
    "IgnoredTables": [],
    "IgnoredColumns": [],
    "MaxExpectedDowntime": "1m"
    "Concurrency": 4
  }
}
```

copydb doesn't require changes because previously it didn't allow for verifier configurations.

Future improvements
-------------------

One issue that exists is the table schema cache must be given to the verifiers after Ferry.Start. This causes some ugly code (Verifier.SetApplicableTableSchemaCache) for now. Once we refactor how the TableSchemaCache is handled, this method should go away.

Another issue is that the VerifyBeforeCutover is no longer controllable by the application level. The application level doesn't even really get an update about this development. This can be addressed in  the future.

### CODE REVIEW

File: compression_verifier.go [hunk 1]
```diff
@@ -21,12 +21,10 @@ const (
 )
 
 type (
-	columnCompressionConfig map[string]string
```
Comment (shuhaowu) [inline compression_verifier.go]: Removed this type because it is only used once and it's also now a part of the public API.

File: ferry.go [hunk 2]
```diff
@@ -288,6 +360,26 @@ func (f *Ferry) Initialize() (err error) {
 	f.DataIterator = f.NewDataIterator()
 	f.BatchWriter = f.NewBatchWriter()
 
+	if f.Config.VerifierType != "" {
+		if f.Verifier != nil {
+			return errors.New("VerifierType specified and Verifier is given. these are mutually exclusive options")
```
Comment (hkdsun) [inline ferry.go]: Why do we support both? Wouldn't it be simpler to have a single codepath?

File: data_iterator.go [hunk 3]
```diff
@@ -25,7 +24,7 @@ type DataIterator struct {
 	logger         *logrus.Entry
 }
 
-func (d *DataIterator) Run() {
+func (d *DataIterator) Run(tables []*schema.Table) {
```
Comment (hkdsun) [inline data_iterator.go]: Why this change?

File: sharding/test/compression_verifier_test.go [hunk 4]
```diff
@@ -67,6 +67,13 @@ func (t *CompressionVerifierTestSuite) InsertCompressedRowInDb(id int, schema, d
 	t.Require().Nil(err)
 }
 
+func (t *CompressionVerifierTestSuite) replaceCompressionVerifier(tableCompressions ghostferry.TableColumnCompressionConfig) {
+	var err error
+	iterativeVerifier := t.Ferry.Ferry.Verifier.(*ghostferry.IterativeVerifier)
+	iterativeVerifier.CompressionVerifier, err = ghostferry.NewCompressionVerifier(tableCompressions)
+	t.Require().Nil(err)
+}
```
Comment (hkdsun) [inline sharding/test/compression_verifier_test.go]: This is..really ugly. Can you do any better with some more refactoring?

File: iterative_verifier.go [hunk 5]
```diff
@@ -244,6 +245,11 @@ func (v *IterativeVerifier) VerifyBeforeCutover() error {
 	return err
 }
 
+func (v *IterativeVerifier) SetApplicableTableSchemaCache(t TableSchemaCache) {
+	v.Tables = t.AsSlice()
+	v.TableSchemaCache = t
+}
```
Comment (hkdsun) [inline iterative_verifier.go]: I still don't get why you're doing this instead of creating the verifier with the right fields

File: ferry.go [hunk 6]
```diff
@@ -344,7 +436,12 @@ func (f *Ferry) Start() error {
 
 	// TODO(pushrax): handle changes to schema during copying and clean this up.
 	f.BinlogStreamer.TableSchema = f.Tables
-	f.DataIterator.Tables = f.Tables.AsSlice()
+
+	// TODO: refactor the TableSchemaCache logic to remove this weird post initialize
+	//       set tables hack
+	if f.Verifier != nil {
+		f.Verifier.SetApplicableTableSchemaCache(f.Tables)
+	}
```
[the remark this answers is not in the document]
Comment (shuhaowu) [reply ferry.go]: f.Tables is only loaded in `Start`. I tried to move this to initialize, but I have to change the integration tests quite a bit as the integration tests depends the tables being loaded in Start.

I started this work but I took it out because it confused this PR too much. I'll create another PR to fix this after merging this one.

File: ferry.go (hunk 2 above)
Comment (shuhaowu) [reply ferry.go]: I only have this because `Verifier` is an `interface`. While we don't do it in our use cases, you could conceivably configure and pass in your own verifier to the Ferry without having the Ferry initialize it for you. You could even create an entirely different verifier without having to change the Ferry.

The code after this doesn't really change with respect to how the verifier is initialized, so the only "complexity" should be the few lines of code here.

File: data_iterator.go (hunk 3 above)
Comment (shuhaowu) [reply data_iterator.go]: I wanted to refactor the table loading logic from `Start` to `Initialize`. I thought I took this part out, but I guess I didn't do it fully. 

As stated [earlier](https://github.com/Shopify/ghostferry/pull/89#discussion_r259859369), I'll refactor this part anyway, so I'm not going to remove it from this PR anymore.

File: iterative_verifier.go (hunk 5 above)
Comment (shuhaowu) [reply iterative_verifier.go]: As stated [earlier](https://github.com/Shopify/ghostferry/pull/89#discussion_r259859369), the loading order for the TableSchemaCache is somewhat important to this. I want to get rid of this too but it proved too difficult for one PR. There'll be a follow up PR that removes this method.

File: sharding/test/compression_verifier_test.go (hunk 4 above)
Comment (shuhaowu) [reply sharding/test/compression_verifier_test.go]: I just tried and I can't do this without major rework of this test. At that point, it would be nice to be able to extract these tests out to the ruby tests, where we can more easily handle configuration changes.

Between this and the issue with table loading in the Go integration tests, we might want to think about refactoring all the tests into the ruby tests. However, I'm not sure if we'll have time to do this for now.

**Comment by shuhaowu** [comment]
As stated, I'll have a follow up PR to cleanup the table loading logic. I'm uncertain about how we want to deal with the compression tests, but lets talk about it offline.

I'll merge this for now so we can have a new base to work on.
